clippy_lints/src/no_mangle_with_rust_abi.rs

```diff
@@ -37,33 +46,75 @@ declare_lint_pass!(NoMangleWithRustAbi => [NO_MANGLE_WITH_RUST_ABI]);
-                    && !fn_attrs.contains("extern")
-                {
-                    let sugg_span = fn_sig.span
-                        .with_lo(fn_sig.span.lo() + BytePos::from_usize(fn_attrs.len()))
-                        .shrink_to_lo();
+        if cx.tcx.get_attr(item.owner_id, sym::no_mangle).is_none() {
+            return;
+        }
 
-                    span_lint_and_then(
-                        cx,
-                        NO_MANGLE_WITH_RUST_ABI,
-                        fn_sig.span,
-                        "`#[no_mangle]` set on a function with the default (`Rust`) ABI",
-                        |diag| {
-                            diag.span_suggestion(sugg_span, "set an ABI", "extern \"C\" ", app)
-                                .span_suggestion(sugg_span, "or explicitly set the default", "extern \"Rust\" ", app);
-                        },
-                    );
-                }
-            }
+        match item.kind {
+            ItemKind::Fn(fn_sig, _, _) => check_fn(cx, fn_sig),
+            ItemKind::Static(ty, _, _) => check_static(cx, item, ty),
+            _ => {},
         }
     }
 }
+
+/// Check for functions that are implicitly using the Rust ABI.
+fn check_fn(cx: &LateContext<'_>, fn_sig: FnSig<'_>) {
+    if fn_sig.header.abi == Abi::Rust
+        && let Some(snippet) = snippet_opt(cx, fn_sig.span)
+        && let Some((fn_attrs, _)) = snippet.split_once("fn")
+        && !fn_attrs.contains("extern")
+    {
+        let sugg_span = fn_sig.span
+            .with_lo(fn_sig.span.lo() + BytePos::from_usize(fn_attrs.len()))
+            .shrink_to_lo();
+
+        span_lint_and_then(
+            cx,
+            NO_MANGLE_WITH_RUST_ABI,
+            fn_sig.span,
+            "`#[no_mangle]` set on a function with the default (`Rust`) ABI",
+            |diag| {
+                diag.span_suggestion(
+                    sugg_span,
+                    "set an ABI",
+                    "extern \"C\" ",
+                    Applicability::MaybeIncorrect,
+                )
+                .span_suggestion(
+                    sugg_span,
+                    "or explicitly set the default",
+                    "extern \"Rust\" ",
+                    Applicability::MaybeIncorrect,
+                );
+            },
+        );
+    }
+}
+
+/// Check for static items with a type that is implicitly using the Rust ABI.
+fn check_static(cx: &LateContext<'_>, item: &Item<'_>, ty: &Ty<'_>) {
+    // TODO(Centri3): Once/if Rust allows an explicit `#[repr(Rust)]`, change this to check for
+    // that. It should work by checking `ReprOptions::flags` for `IS_EXPLICIT_RUST`, whatever
+    // value that may be.
+
+    if let ty::Adt(def, _) = hir_ty_to_ty(cx.tcx, ty).kind()
+        // No explicit representation. `align` and `pack` don't need to be checked as,
+        // afaik, what mostly matters is just the Rust ABI.
+        && def.repr().flags.bits() == 0
```

Comment:
Checking the flags directly runs into some implementation details, e.g. if `-Z randomize-layout` is passed the lint won't fire

Let's go with a more high level `!repr.c() && !repr.transparent() && !repr.simd() && repr.int.is_none()`. The latter because `#[repr(int)]` on enums have defined layout

Reply:
Out of curiosity, does `#[repr(int)]` define a variant's layout if it's a struct/tuple variant? Or is that what `#[repr(C, int)]` does? Because if so, we should probably lint that too if there are any variants that aren't unit.

Reply:
Yep - https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md#guide-level-explanation, you don't have to check the variant kinds